In [1]:
import pandas as pd
import math

In [2]:
pandas_df = pd.read_csv('../data/input/IMDB-Movie-Data.csv')
print(pandas_df.info())
pandas_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB
None


Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

#### Challenge 1. Warm up
We want to create bins of movies according to the number of votes they've received. For that matter, we will create a new column named 'bin' which will tag every movie as follow:

From 0 to 999 ==> 1  
From 1000 to 9999 ==> 2  
From 10000 to 99999 ==> 3  
From 100000 to 999999 ==> 4  
More than 1000000 ==> 5

In [3]:
pandas_df['bins'] = pandas_df['Votes'].apply(lambda x: 1 if x < 1000 else 2 if x < 10000 else 3 if x < 100000 else 4 if x < 1000000 else 5)
pandas_df.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  bins  
0     8.1  757074              333.13       76.0     4  
1     7.0  485820              126.46       65.0     4  
2     7.3  157606              138.12       62.0     4  
3     7.2   60545              270.32       59.0     3  
4     6.2  393727              325.02       40.0     4

**Challenge 2. Using axis concept**
We want to know how much is the revenue per minute for every movie.

In [4]:
pandas_df['Revenue per Minute'] = pandas_df.apply(lambda x: x['Revenue (Millions)'] / x['Runtime (Minutes)'], axis=1)
pandas_df.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  bins  Revenue per Minute  
0     8.1  757074              333.13       76.0     4            2.753140  
1     7.0  485820              126.46       65.0     4            1.019839  
2     7.3  157606              138.12       62.0     4            1.180513  
3     7.2   60545              270.32       59.0     3            2.502963  
4     6.2  393727              325.02       40.0     4            2.642439

**Challenge 3. Using the lambda**
We want to create a new ranking where we add 1 point if the genre is thriller but subtract 1 point if the genre is comedy.

In [5]:
pandas_df['New Ranking'] = pandas_df.apply(lambda x: x['Rank'] + 1 if 'Thriller' in x['Genre'] else x['Rank'] - 1 if 'Comedy' in x['Genre'] else x['Rank'], axis=1)
pandas_df.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  bins  Revenue per Minute  \
0     8.1  757074              333.13       76.0     4            2.753140   
1     7.0  485820              126.46       65.0     4            1.019839   
2     7.3  157606              138.12       62.0     4            1.180513   
3     7.2   60545              270.32       59.0     3            2.502963   
4     6.2  393727              325.02       40.0     4            2.642439   

   New Ranking  
0            1  
1            2  
2            4  
3            3  
4            5

**Challenge 4. Now the real stuff**
We want to know if the sum of the ASCII value of every character of the movie title divided by the number of votes retrieve a prime number...remember, prime numbers are integers.

In [6]:
ascii_sum = lambda x: sum(ord(c) for c in x['Title']) / x['Votes']

result = pandas_df.apply(ascii_sum, axis=1)
result

0      0.002866
1      0.002198
2      0.003325
3      0.006623
4      0.003180
         ...   
995    0.066667
996    0.017744
997    0.026592
998    0.237246
999    0.075674
Length: 1000, dtype: float64

**Challenge 5. And finally some fantasy**
Feel free to propose your own ranking based in aggregations of at least 3 columns of the dataset.

In [13]:
pandas_df['My Ranking'] = pandas_df.apply(lambda x: x['Rank'] +3 if 'Comedy' in x['Genre'] \
                                          else x['Rank'] +3 if 'Action' in x['Genre'] \
                                          else x['Rank']+5 if x['Year'] > 2022 \
                                          else x['Rank']+3 if x['Runtime (Minutes)']< 90 \
                                          else "Not Good", axis=1)
pandas_df.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  bins  Revenue per Minute  \
0     8.1  757074              333.13       76.0     4            2.753140   
1     7.0  485820              126.46       65.0     4            1.019839   
2     7.3  157606              138.12       62.0     4            1.180513   
3     7.2   60545              270.32       59.0     3            2.502963   
4     6.2  393727              325.02       40.0     4            2.642439   

   New Ranking My Ranking  
0            1          4  
1            2   Not Good  
2            4   Not Good  
3            3          7  
4            5          8

**Challenge 6. Freaky bonus**
We want to know which movies might have hidden paterns in their description. A way to know that is finding those movies which the sum of all numeric values of the string description hash (SHA256) are between their revenue and their number of votes.